### PR TITLE
soc: sam0: dynamically enable the SAM0 drivers

### PR DESCRIPTION
**Pull request by nzmichaelh**
soc: sam0: dynamically enable the SAM0 drivers

Change the SAM0 to match other boards by selecting the SAM0 specific driver when a driver class is selected.

For example, automatically enable CONFIG_SPI_SAM0 when CONFIG_SPI is enabled.

Change the Arduino Zero to match.  This reduces the size of `blink` from ~14 KiB to ~10 KiB.

Also add the supported features to arduino_zero.yaml to get more coverage in the sanity tests.

### CODE REVIEW

File: boards/arm/arduino_zero/Kconfig.defconfig
Comment (galak) [inline boards/arm/arduino_zero/Kconfig.defconfig:11]: Remove `SPI_4` here, this should have no impact as the only driver `SPI_4` has meaning on is the nrfx SPI driver.

File: soc/arm/atmel_sam0/common/Kconfig.defconfig.series
Comment (galak) [inline soc/arm/atmel_sam0/common/Kconfig.defconfig.series:56]: Fix comment here. SOC_FAMILY_SAMD21 -> SOC_FAMILY_SAM0

**Review by galak — CHANGES_REQUESTED** [review]
In general looks good, two minor tweaks.

File: boards/arm/arduino_zero/Kconfig.defconfig
Comment (nzmichaelh) [reply boards/arm/arduino_zero/Kconfig.defconfig:11]: Done

File: soc/arm/atmel_sam0/common/Kconfig.defconfig.series
Comment (nzmichaelh) [reply soc/arm/atmel_sam0/common/Kconfig.defconfig.series]: Done

**Comment by nzmichaelh** [comment]
I've removed the entropy driver hook as the SAMD21 doesn't have the TRNG.

**Review by galak — APPROVED** [review]

**Comment by nzmichaelh** [comment]
I've added back in PINMUX to the Arduino Zero defaults, as else nothing works as the pins aren't configured.

**Review by stephanosio — COMMENTED** [review]
Looks ok to me; but, this change arguably goes against the the long discussion we had in #24561, so I will leave it up to the Cortex-M maintainer (@ioannisg).

File: soc/arm/atmel_sam0/common/Kconfig.defconfig.series
Comment (carlescufi) [inline soc/arm/atmel_sam0/common/Kconfig.defconfig.series:9]: I think this would be cleaner if you did:

```
config ADC_SAM0
    default ADC
```

This means that `ADC_SAM0` inherits the `y/n` value from `ADC`.
Same for all others.

@MaureenHelm @galak Did we settle on using this idiom? I thought so at least

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Comment by carlescufi** [comment]
> Looks ok to me; but, this change arguably goes against the the long discussion we had in #24561, so I will leave it up to the Cortex-M maintainer (@ioannisg).

I don't believe this applies to peripherals @stephanosio, only to very specific configuration options like MPU. Otherwise we end up with massive board kconfig files. Feedback welcome @galak @ioannisg  @MaureenHelm

File: soc/arm/atmel_sam0/common/Kconfig.defconfig.series
Comment (nzmichaelh) [reply soc/arm/atmel_sam0/common/Kconfig.defconfig.series:9]: Done

**Review by MaureenHelm — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
Thank you very much @nzmichaelh for following-up on this and for your patience.

**Pull request merged**
